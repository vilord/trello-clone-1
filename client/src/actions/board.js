import * as types from '../constants/actionTypes';

export const setBoardTitle = title => ({
  type: types.SET_BOARD_TITLE,
  title,
});

export const getBoardRequest = () => ({
  type: types.GET_BOARD_REQUEST,
});

export const getBoardSucess = board => ({
  type: types.GET_BOARD_SUCESS,
  board,
});

export const getBoardFailure = () => ({
  type: types.GET_BOARD_FAILURE,
});

export const getBoard = id => async dispatch => {
  dispatch(getBoardRequest());
  try {
    const res = await fetch(`/boards/${id}`, {
      credentials: 'include',
      headers: {
        'content-type': 'application/json',
      },
    });

    if (res.status === 200) {
      const { board } = await res.json();
      return dispatch(getBoardSucess(board));
    }

    dispatch(getBoardFailure());
  } catch (err) {
    dispatch(getBoardFailure());
    console.log(err);
  }
};
