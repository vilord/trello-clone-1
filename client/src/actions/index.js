export const INFO_REQ = 'INFO_REQ';
export const TOGGLE_SHOW = 'TOGGLE_SHOW';
export const REQUEST_POSTS = 'REQUEST_POSTS';
export const RECEIVE_POSTS = 'RECEIVE_POSTS';

export const toggleShow = {
  type: TOGGLE_SHOW,
};

export const requestPosts = subreddit => ({
  type: REQUEST_POSTS,
  subreddit,
});

export const receivePosts = (subreddit, data) => ({
  type: RECEIVE_POSTS,
  subreddit,
  data,
});

/**
 * Async Actions
 */
export const fetchPosts = subreddit => {
  return dispatch => {
    dispatch(requestPosts(subreddit));
    return fetch(`https://www.reddit.com/r/${subreddit}.json`)
      .then(
        response => response.json(),
        error => console.log('An error occurred.', error),
      )
      .then(data => dispatch(receivePosts(subreddit, data)));
  };
};
