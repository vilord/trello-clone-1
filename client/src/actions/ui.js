import * as types from '../constants/actionTypes';

/**
 * UI Errors
 */
export const setUiError = error => ({
  type: types.SET_UI_ERROR,
  error,
});

export const resetUiError = () => ({
  type: types.RESET_UI_ERROR,
});

/**
 * Boards Explorer
 */
export const showBoardsExplorer = () => ({
  type: types.SHOW_BOARDS_EXPLORER,
});

export const hideBoardsExplorer = () => ({
  type: types.HIDE_BOARDS_EXPLORER,
});

/**
 * Header Search
 */
export const focusHeaderSearch = () => ({
  type: types.FOCUS_HEADER_SEARCH,
});

export const blurHeaderSearch = () => ({
  type: types.BLUR_HEADER_SEARCH,
});

export const setHeaderSearch = term => ({
  type: types.SET_HEADER_SEARCH,
  term,
});

/**
 * Create Menu
 */
export const showCreateMenu = () => ({
  type: types.SHOW_CREATE_MENU,
});

export const hideCreateMenu = () => ({
  type: types.HIDE_CREATE_MENU,
});

/**
 * Notifications
 */
export const showNotifications = () => ({
  type: types.SHOW_NOTIFICATIONS,
});

export const hideNotifications = () => ({
  type: types.HIDE_NOTIFICATIONS,
});

/**
 * User Menu
 */
export const showUserMenu = () => ({
  type: types.SHOW_USER_MENU,
});

export const hideUserMenu = () => ({
  type: types.HIDE_USER_MENU,
});

/**
 * Rename Board
 */
export const showRenameBoard = () => ({
  type: types.SHOW_RENAME_BOARD,
});

export const hideRenameBoard = () => ({
  type: types.HIDE_RENAME_BOARD,
});

/**
 * Add to Team
 */
export const showAddToTeam = () => ({
  type: types.SHOW_ADD_TO_TEAM,
});

export const hideAddToTeam = () => ({
  type: types.HIDE_ADD_TO_TEAM,
});

/**
 * Visibility Menu
 */
export const showVisibilityMenu = () => ({
  type: types.SHOW_VISIBILITY_MENU,
});

export const hideVisibilityMenu = () => ({
  type: types.HIDE_VISIBILITY_MENU,
});

/**
 * Board Menu
 */
export const showBoardMenu = () => ({
  type: types.SHOW_BOARD_MENU,
});

export const hideBoardMenu = () => ({
  type: types.HIDE_BOARD_MENU,
});
