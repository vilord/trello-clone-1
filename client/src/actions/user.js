import * as types from '../constants/actionTypes';
import * as errors from '../constants/errors';

/*
 * Signup User
 */
export const signupUserRequest = () => ({
  type: types.SIGNUP_USER_REQUEST,
});

export const signupUserSuccess = user => ({
  type: types.SIGNUP_USER_SUCCESS,
  user,
});

export const signupUserFailure = error => ({
  type: types.SIGNUP_USER_FAILURE,
  error,
});

/**
 * Login User
 */
export const loginUserRequest = () => ({
  type: types.LOGIN_USER_REQUEST,
});

export const loginUserSuccess = user => ({
  type: types.LOGIN_USER_SUCCESS,
  user,
});

export const loginUserFailure = error => ({
  type: types.LOGIN_USER_FAILURE,
  error,
});

/**
 * Logout User
 */
export const logoutUser = () => ({
  type: types.LOGOUT_USER,
});

/**
 * Get User Session
 */
export const getUserSessionRequest = () => ({
  type: types.GET_USER_SESSION_REQUEST,
});

export const getUserSessionSuccess = user => ({
  type: types.GET_USER_SESSION_SUCCESS,
  user,
});

/**
 * Set User Profile
 */
export const setUserProfileRequest = () => ({
  type: types.SET_USER_PROFILE_REQUEST,
});

export const setUserProfileSuccess = profile => ({
  type: types.SET_USER_PROFILE_SUCCESS,
  profile,
});

export const setUserProfileFailure = error => ({
  type: types.SET_USER_PROFILE_FAILURE,
  error,
});

/**
 * Async Actions
 */
export const signupUser = (newUser, history) => async dispatch => {
  dispatch(signupUserRequest());

  try {
    const res = await fetch(`/users`, {
      method: 'POST',
      credentials: 'include',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify(newUser),
    });

    const { user, error } = await res.json();

    if (res.status === 200 && user) {
      dispatch(signupUserSuccess(user));
      history.push('/');
      return;
    }

    if (res.status === 409 && error) {
      return dispatch(signupUserFailure(errors.userExists));
    }

    return dispatch(signupUserFailure(errors.serverError));
  } catch (err) {
    dispatch(signupUserFailure(errors.serverError));
    console.log(err);
  }
};

export const loginUser = (userLoginInfo, history) => async dispatch => {
  dispatch(loginUserRequest());

  try {
    const res = await fetch(`/auth/login`, {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify(userLoginInfo),
      credentials: 'include',
    });

    if (res.status === 200) {
      const { user } = await res.json();
      dispatch(loginUserSuccess(user));
      history.push('/');
      return;
    }

    dispatch(loginUserFailure(errors.wrongCredentials));
  } catch (err) {
    console.log(err);
    dispatch(loginUserFailure(errors.serverError));
  }
};

export const getUserSession = history => async dispatch => {
  dispatch(getUserSessionRequest());

  try {
    const res = await fetch('/auth/user-session', {
      credentials: 'include',
      headers: {
        'content-type': 'application/json',
      },
    });

    const { user } = await res.json();

    if (res.status === 200 && user) {
      dispatch(getUserSessionSuccess(user));
      history.push('/');
      return;
    }

    dispatch(logoutUser());
    history.push('/login');
  } catch (err) {
    console.log(err);
    dispatch(logoutUser());
  }
};

export const sendLogoutUser = () => async dispatch => {
  try {
    const res = await fetch('/auth/logout', {
      method: 'POST',
      credentials: 'inlude',
    });

    if (res.status !== 200) throw new Error(await res.json());
    dispatch(logoutUser());
  } catch (err) {
    console.log(err);
    dispatch(logoutUser());
  }
};

export const setUserProfile = newProfile => async dispatch => {
  dispatch(setUserProfileRequest());

  try {
    const res = await fetch('/users/profile', {
      method: 'PUT',
      credentials: 'include',
      body: {
        profile: newProfile,
      },
    });

    const { profile, error } = await res.json();

    if (res.status === 200 && profile) {
      return dispatch(setUserProfileSuccess(profile));
    }

    return dispatch(setUserProfileFailure(error));
  } catch (err) {
    console.log(err);
  }
};
