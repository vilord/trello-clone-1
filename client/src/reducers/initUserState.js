export default {
  name: '',
  username: '',
  email: '',
  bio: '',
  avatar: '',
  boards: [],
  teams: [],
  assigned_cards: [],
  activity: [],
};
