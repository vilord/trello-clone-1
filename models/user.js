const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const { ObjectId } = Schema.Types;
const isEmail = require('validator/lib/isEmail');
const isUrl = require('validator/lib/isUrl');

const UserSchema = new Schema(
  {
    username: {
      type: String,
      unique: true,
      trim: true,
      match: /^((?!\s+).)*$/,
      required: true,
    },
    password: {
      type: String,
      minlength: 6,
      maxlength: 256,
      // TODO: password match
      // match: /^.*[a-zA-Z].*[0-9]+.*$/,
      bcrypt: true,
      required: function() {
        return !this.google_id;
      },
    },
    google_id: {
      type: Number,
      required: function() {
        return !this.password;
      },
    },
    email: {
      type: String,
      unique: true,
      trim: true,
      lowercase: true,
      required: true,
      validate: {
        validator: x => isEmail(x),
        message: 'Must be a valid email address',
      },
    },
    name: {
      type: String,
      required: true,
      trim: true,
      match: /^((?!\s{2,}).)*$/,
      maxlength: 70,
    },
    initials: {
      type: String,
      uppercase: true,
      maxlength: 4,
      minlength: 1,
      match: /^((?!\s).)*$/,
    },
    bio: {
      type: String,
      trim: true,
      minlength: 1,
      maxlength: 500,
    },
    avatar: {
      type: String,
      trim: true,
      validate: {
        validator: x => isUrl(x),
        message: 'it needs a valid URL',
      },
    },
    assigned_cards: [
      {
        type: ObjectId,
        ref: 'Card',
      },
    ],
    boards: [
      {
        board: {
          type: ObjectId,
          ref: 'Board',
        },
        favorite: {
          type: Boolean,
          default: false,
        },
      },
    ],
    teams: [
      {
        type: ObjectId,
        ref: 'Team',
      },
    ],
    activity: [
      {
        type: ObjectId,
        ref: 'Activity',
      },
    ],
  },
  {
    timestamps: true,
  },
);

/**
 * Password Hashing and Salting Plugin
 */
UserSchema.plugin(require('mongoose-bcrypt'), { rounds: 8 });

/**
 * Remove sensitive information by default
 * when transforming the model toObject.
 */
UserSchema.options.toObject = {
  transform: function(doc, ret, options) {
    delete ret._id;
    delete ret.__v;
    delete ret.password;
    delete ret.createdAt;
    delete ret.updatedAt;
    delete ret.google_id;
    return ret;
  },
};

/**
 * Pre validation hook
 */
UserSchema.pre('validate', async function(next) {
  if (!this.name) {
    const err = new Error('User validation failed: name is required.');
    err.name = 'ValidationError';
    return next(err);
  }

  // Generates unique username based on the name.
  const genUsername = async name => {
    let username = name.replace(/\s/g, '').toLowerCase();
    const count = await this.model('User')
      .find({ username: { $regex: username } })
      .lean()
      .count();
    username += count === 0 ? '' : count;
    return username;
  };

  this.username = this.username || (await genUsername(this.name));

  next();
});

module.exports = mongoose.model('User', UserSchema);
