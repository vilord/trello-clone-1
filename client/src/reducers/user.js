import * as types from '../constants/actionTypes';
import initUserState from './initUserState';

const userReducer = (state = initUserState, action) => {
  switch (action.type) {
    case types.SIGNUP_USER_SUCCESS:
      return {
        ...state,
        ...action.user,
      };
    case types.GET_USER_SESSION_SUCCESS:
    case types.LOGIN_USER_SUCCESS:
      return {
        ...state,
        ...action.user,
      };
    case types.SET_USER_PROFILE_SUCCESS:
      return {
        ...state,
        ...action.profile,
      };
    default:
      return state;
  }
};

/**
 * Higher order function that resets
 * the state to its initial value on
 * LOGOUT_USER action.
 */
const logoutWrapper = (state, action) => {
  if (action.type === types.LOGOUT_USER) {
    state = undefined;
  }
  return userReducer(state, action);
};

export default logoutWrapper;
