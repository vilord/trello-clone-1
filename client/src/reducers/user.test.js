import userReducer from './user';
import initUserState from './initUserState';

import * as actions from '../actions/user';
// import * as types from '../constants/actionTypes';

describe('User Reducer', () => {
  beforeAll(() => {
    console.error = err => {
      throw new Error(err);
    };
  });

  let initState;

  beforeEach(() => {
    initState = { ...initUserState };
  });

  it('returns init state when state arg is undefined', () => {
    expect(userReducer(undefined, {})).toEqual(initState);
  });

  it('handles LOGIN_USER_SUCCESS', () => {
    const user = {
      name: 'John Doe',
      username: 'johndoe',
      initials: 'JD',
      bio: 'Web Developer',
      avatar: 'http://example.com/prfile_pic',
    };

    const nextState = {
      ...initState,
      ...user,
    };

    expect(userReducer(initState, actions.loginUserSuccess(user))).toEqual(
      nextState,
    );
    expect(
      userReducer(initState, actions.loginUserSuccess(user)),
    ).toMatchSnapshot();
  });

  it('handles LOGOUT_USER', () => {
    const currState = {
      name: 'John Doe',
      username: 'johndoe',
    };

    expect(userReducer(currState, actions.logoutUser())).toEqual(initState);
    expect(userReducer(currState, actions.logoutUser())).toMatchSnapshot();
  });

  it('handles SET_USER_PROFILE_SUCCESS', () => {
    initState = {
      ...initState,
      name: 'John Doe',
      username: 'johndoe',
    };

    const profile = {
      name: 'John Doe Simpson',
      username: 'johndoes',
      initials: 'JDS',
    };

    const nextState = {
      ...initState,
      ...profile,
    };

    expect(
      userReducer(initState, actions.setUserProfileSuccess(profile)),
    ).toEqual(nextState);
    expect(
      userReducer(initState, actions.setUserProfileSuccess(profile)),
    ).toMatchSnapshot();
  });
});
