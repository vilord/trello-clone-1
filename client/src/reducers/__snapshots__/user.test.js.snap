// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Reducer handles LOGIN_USER_SUCCESS 1`] = `
Object {
  "activity": Array [],
  "assigned_cards": Array [],
  "avatar": "http://example.com/prfile_pic",
  "bio": "Web Developer",
  "boards": Array [],
  "email": "",
  "initials": "JD",
  "name": "John Doe",
  "teams": Array [],
  "username": "johndoe",
}
`;

exports[`User Reducer handles LOGOUT_USER 1`] = `
Object {
  "activity": Array [],
  "assigned_cards": Array [],
  "avatar": "",
  "bio": "",
  "boards": Array [],
  "email": "",
  "name": "",
  "teams": Array [],
  "username": "",
}
`;

exports[`User Reducer handles SET_USER_PROFILE_SUCCESS 1`] = `
Object {
  "activity": Array [],
  "assigned_cards": Array [],
  "avatar": "",
  "bio": "",
  "boards": Array [],
  "email": "",
  "initials": "JDS",
  "name": "John Doe Simpson",
  "teams": Array [],
  "username": "johndoes",
}
`;
